#!/usr/bin/env nextflow
/*
========================================================================================
                         sheynkman-lab/Long-Read-Proteogenomics
========================================================================================
 sheynkman-lab/Long-Read-Proteogenomics Analysis Pipeline.
 #### Homepage / Documentation
 https://github.com/sheynkman-lab/Long-Read-Proteogenomics
----------------------------------------------------------------------------------------
*/

def helpMessage() {
    // TODO nf-core: Add to this help message with new command line parameters
    log.info logHeader()
    log.info"""

    Usage:

    The typical command for running the pipeline is as follows:

    nextflow run sheynkman-lab/Long-Read-Proteogenomics --input -profile docker

    Mandatory arguments:
      --input [file]                  Path to input data (must be surrounded with quotes)
      -profile [str]                  Configuration profile to use. Can use multiple (comma separated)
                                      Available: conda, docker, singularity, test, awsbatch, <institute> and more

    Options:
      --single_end [bool]             Specifies that the input is single-end reads

    References                        If not specified in the configuration file or you wish to overwrite any of the references
      --fasta [file]                  Path to fasta reference

    Other options:
      --outdir [file]                 The output directory where the results will be saved
      --publish_dir_mode [str]        Mode for publishing results in the output directory. Available: symlink, rellink, link, copy, copyNoFollow, move (Default: copy)
      --email [email]                 Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits
      --email_on_fail [email]         Same as --email, except only send mail if the workflow is not successful
      --max_multiqc_email_size [str]  Threshold size for MultiQC report to be attached in notification email. If file generated by pipeline exceeds the threshold, it will not be attached (Default: 25MB)
      -name [str]                     Name for the pipeline run. If not specified, Nextflow will automatically generate a random mnemonic

    AWSBatch options:
      --awsqueue [str]                The AWSBatch JobQueue that needs to be set when running on AWSBatch
      --awsregion [str]               The AWS Region for your AWS Batch job to run on
      --awscli [str]                  Path to the AWS CLI tool
    """.stripIndent()
}

// Show help message
if (params.help) {
    helpMessage()
    exit 0
}

/*
 * SET UP CONFIGURATION VARIABLES
 */


// Header log info
log.info logHeader()
def summary = [:]
if (workflow.revision) summary['Pipeline Release'] = workflow.revision
// TODO nf-core: Report custom parameters here
if (params.input) summary['Input']            = params.input
summary['Max Resources']    = "$params.max_memory memory, $params.max_cpus cpus, $params.max_time time per job"
summary['Output dir']       = params.outdir
if (workflow.profile.contains('awsbatch')) {
    summary['AWS Region']   = params.awsregion
    summary['AWS Queue']    = params.awsqueue
    summary['AWS CLI']      = params.awscli
}
summary['Config Profile'] = workflow.profile
log.info summary.collect { k,v -> "${k.padRight(18)}: $v" }.join("\n")
log.info "-\033[2m--------------------------------------------------\033[0m-"

/*
 * STEP  - validate template
 */

ch_raw_pacbio_bams_folder = params.raw_pacbio_bams_folder ? Channel.fromPath("${params.raw_pacbio_bams_folder}/*.{bam,bam.bai}") : null
ch_design_reads_csv = params.input ? Channel.fromPath("${params.input}") : null

// Fail early: Do not allow the user to provide input file and input folder simultaneously
if (params.input && params.raw_pacbio_bams_folder ) {
    exit 1, "You cannot provide an input file and input folder simultaneously. File: ${params.input}\n, Folder: ${params.raw_pacbio_bams_folder},\nSee --help for more information"
}

// Fail early: Do not allow the user to provide neither input file nor input folder
if (!params.input && !params.raw_pacbio_bams_folder ) {
    exit 1, "Malformed row in TSV file: ${row}, see --help for more information"
}

// If the user has provided input folder
if (!params.input && params.raw_pacbio_bams_folder ) {
    ch_raw_pacbio_bams_folder
        .map { it -> [ it ] }
        .set { ch_raw_pacbio_subreads_bams }
}

// // If the user has provided input design file
// if (params.input && !params.raw_pacbio_bams_folder ) {
//     ch_design_reads_csv
//         .splitCsv(header:true, sep:',')
//         .map { row -> [ file(row).simpleName, bam, bai ] }
//         .into { ch_raw_pacbio_subreads_bams }
// }

ch_raw_pacbio_subreads_bams.view()


/*
 * STEP  - validate template
 */

process validate {
    publishDir "${params.outdir}/validate/", mode: params.publish_dir_mode

    output:
    file "validated.txt"

    script:
    """
    touch validated.txt
    """
}

/*
 *  Module 1: SMARTLink - CCS
 */

process smartlink_ccs {
    tag "${sample}"
    publishDir "${params.outdir}/smartlink_ccs/", mode: params.publish_dir_mode

    input:
    set val(sample), file(raw_pacbio_subreads_bam), file(raw_pacbio_subreads_bai) from ch_raw_pacbio_subreads_bams

    output:
    set val("${sample}"), file("${sample}*bam"), file("${sample}*bai") into ch_ccs_pacbio_bams

    script:
    """
    echo "when in pairs:"
    echo "simpleName:${sample}\nbam:${raw_pacbio_subreads_bam}\nbai:${raw_pacbio_subreads_bai}"
    echo "simpleName:${sample}\nbam:${raw_pacbio_subreads_bam}\nbai:${raw_pacbio_subreads_bai}" > "${sample}_fake_input.txt"
    """
}

process isoseq3 {
    tag "${sample}"
    publishDir "${params.outdir}/isoseq3/", mode: params.publish_dir_mode

    input:
    set val("${sample}"), file("${sample}*bam"), file(bai) from ch_ccs_pacbio_bams

    output:
    file("*completed.bam")

    script:
    """
    echo "when in pairs:"
    echo "simpleName:${sample}\nbam:${raw_pacbio_subreads_bam}\nbai:${raw_pacbio_subreads_bai}"
    echo "simpleName:${sample}\nbam:${raw_pacbio_subreads_bam}\nbai:${raw_pacbio_subreads_bai}" > "${sample}_fake_input.txt"
    """
}







def logHeader() {
    // Log colors ANSI codes
    c_black = params.monochrome_logs ? '' : "\033[0;30m";
    c_blue = params.monochrome_logs ? '' : "\033[0;34m";
    c_cyan = params.monochrome_logs ? '' : "\033[0;36m";
    c_dim = params.monochrome_logs ? '' : "\033[2m";
    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_reset = params.monochrome_logs ? '' : "\033[0m";
    c_white = params.monochrome_logs ? '' : "\033[0;37m";
    c_yellow = params.monochrome_logs ? '' : "\033[0;33m";

    return """    -${c_dim}--------------------------------------------------${c_reset}-
    ${c_cyan}  sheynkman-lab/Long-Read-Proteogenomics v${workflow.manifest.version}${c_reset}
    ${c_cyan}███████╗██╗  ██╗███████╗██╗   ██╗███╗   ██╗██╗  ██╗███╗   ███╗ █████╗ ███╗   ██╗      ██╗      █████╗ ██████╗            ${c_reset}       
    ${c_cyan}██╔════╝██║  ██║██╔════╝╚██╗ ██╔╝████╗  ██║██║ ██╔╝████╗ ████║██╔══██╗████╗  ██║      ██║     ██╔══██╗██╔══██╗           ${c_reset}        
    ${c_cyan}███████╗███████║█████╗   ╚████╔╝ ██╔██╗ ██║█████╔╝ ██╔████╔██║███████║██╔██╗ ██║█████╗██║     ███████║██████╔╝           ${c_reset}
    ${c_cyan}╚════██║██╔══██║██╔══╝    ╚██╔╝  ██║╚██╗██║██╔═██╗ ██║╚██╔╝██║██╔══██║██║╚██╗██║╚════╝██║     ██╔══██║██╔══██╗           ${c_reset}
    ${c_cyan}███████║██║  ██║███████╗   ██║   ██║ ╚████║██║  ██╗██║ ╚═╝ ██║██║  ██║██║ ╚████║      ███████╗██║  ██║██████╔╝           ${c_reset}
    ${c_cyan}╚══════╝╚═╝  ╚═╝╚══════╝   ╚═╝   ╚═╝  ╚═══╝╚═╝  ╚═╝╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝      ╚══════╝╚═╝  ╚═╝╚═════╝            ${c_reset}
    ${c_cyan}                                                                                                                         ${c_reset}
    ${c_cyan}██╗      ██████╗ ███╗   ██╗ ██████╗     ██████╗ ███████╗ █████╗ ██████╗                                                  ${c_reset}
    ${c_cyan}██║     ██╔═══██╗████╗  ██║██╔════╝     ██╔══██╗██╔════╝██╔══██╗██╔══██╗                                                 ${c_reset}
    ${c_cyan}██║     ██║   ██║██╔██╗ ██║██║  ███╗    ██████╔╝█████╗  ███████║██║  ██║                                                 ${c_reset}
    ${c_cyan}██║     ██║   ██║██║╚██╗██║██║   ██║    ██╔══██╗██╔══╝  ██╔══██║██║  ██║                                                 ${c_reset}
    ${c_cyan}███████╗╚██████╔╝██║ ╚████║╚██████╔╝    ██║  ██║███████╗██║  ██║██████╔╝                                                 ${c_reset}
    ${c_cyan}╚══════╝ ╚═════╝ ╚═╝  ╚═══╝ ╚═════╝     ╚═╝  ╚═╝╚══════╝╚═╝  ╚═╝╚═════╝                                                  ${c_reset}
    ${c_cyan}                                                                                                                         ${c_reset}
    ${c_cyan}██████╗ ██████╗  ██████╗ ████████╗███████╗ ██████╗  ██████╗ ███████╗███╗   ██╗ ██████╗ ███╗   ███╗██╗ ██████╗███████╗    ${c_reset}
    ${c_cyan}██╔══██╗██╔══██╗██╔═══██╗╚══██╔══╝██╔════╝██╔═══██╗██╔════╝ ██╔════╝████╗  ██║██╔═══██╗████╗ ████║██║██╔════╝██╔════╝    ${c_reset}
    ${c_cyan}██████╔╝██████╔╝██║   ██║   ██║   █████╗  ██║   ██║██║  ███╗█████╗  ██╔██╗ ██║██║   ██║██╔████╔██║██║██║     ███████╗    ${c_reset}
    ${c_cyan}██╔═══╝ ██╔══██╗██║   ██║   ██║   ██╔══╝  ██║   ██║██║   ██║██╔══╝  ██║╚██╗██║██║   ██║██║╚██╔╝██║██║██║     ╚════██║    ${c_reset}
    ${c_cyan}██║     ██║  ██║╚██████╔╝   ██║   ███████╗╚██████╔╝╚██████╔╝███████╗██║ ╚████║╚██████╔╝██║ ╚═╝ ██║██║╚██████╗███████║    ${c_reset}
    ${c_cyan}╚═╝     ╚═╝  ╚═╝ ╚═════╝    ╚═╝   ╚══════╝ ╚═════╝  ╚═════╝ ╚══════╝╚═╝  ╚═══╝ ╚═════╝ ╚═╝     ╚═╝╚═╝ ╚═════╝╚══════╝    ${c_reset}
    ${c_cyan}                                                                                                                                       
    -${c_dim}--------------------------------------------------${c_reset}-
    """.stripIndent()
}

