{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pacbio = pd.read_csv('../../../data/pacbio_protein_clusters.tsv', sep = '\\t')\n",
    "pacbio['accessions'] = pacbio['pb_accs'].str.split('|')\n",
    "pacbio['base_acc'] = pacbio['accessions'].apply(lambda x : x[0])\n",
    "orf = pd.read_csv('../../../data/orf-testset-fraction16.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_total(accessions, orf_dict):\n",
    "    total = 0\n",
    "    for acc in accessions:\n",
    "        total += orf_dict[acc]\n",
    "    return total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "fl_dict = pd.Series(orf.FL.values,index=orf.pb_acc).to_dict()\n",
    "cpm_dict = pd.Series(orf.CPM.values,index=orf.pb_acc).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pacbio['FL'] = pacbio['accessions'].apply(lambda accs : get_total(accs, fl_dict))\n",
    "pacbio['CPM'] = pacbio['accessions'].apply(lambda accs : get_total(accs, cpm_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = SeqIO.parse(open('../../../data/pacbio_protein_clusters.fasta'), 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../../../data/pacbio_clusters.fasta\", \"w\") as ofile:\n",
    "    for entry in seqs:\n",
    "        seq = str(entry.seq)\n",
    "        pb_acc = entry.id\n",
    "        pb_row = pacbio[pacbio['pb_accs'] == pb_acc].iloc[0]\n",
    "        base_acc = pb_row['base_acc']\n",
    "        gene = orf[orf['pb_acc'] == base_acc].iloc[0]['gene']\n",
    "        ofile.write(f\">pb|{base_acc}|fullname GN={gene}\\n{seq}\\n\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "pacbio = pacbio[['pb_accs', 'base_acc', 'FL', 'CPM']]\n",
    "pacbio.to_csv(\"../../../data/pacbio_clusters.tsv\", sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>pb_accs</th>\n",
       "      <th>accessions</th>\n",
       "      <th>base_acc</th>\n",
       "      <th>FL</th>\n",
       "      <th>CPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MKELGIWEPLAVKLQTYKTAVETAVLLLRIDDIVSGHKKKGDDQSR...</td>\n",
       "      <td>PB.1091.1</td>\n",
       "      <td>[PB.1091.1]</td>\n",
       "      <td>PB.1091.1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.669420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MMLNIINSSITTKAISRWSSLACNIALDAVKMVQFEENGRKEIDIK...</td>\n",
       "      <td>PB.1091.10|PB.1091.19</td>\n",
       "      <td>[PB.1091.10, PB.1091.19]</td>\n",
       "      <td>PB.1091.10</td>\n",
       "      <td>34</td>\n",
       "      <td>11.380147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MMKMLLDPMGGIVMTNDGNAILREIQVQHPAAKSMIEISRTQDEEV...</td>\n",
       "      <td>PB.1091.2|PB.1091.9|PB.1091.11|PB.1091.12|PB.1...</td>\n",
       "      <td>[PB.1091.2, PB.1091.9, PB.1091.11, PB.1091.12,...</td>\n",
       "      <td>PB.1091.2</td>\n",
       "      <td>51</td>\n",
       "      <td>17.070221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MMGHRPVLVLSQNTKRESGRKVQSGNINAAKTIADIIRTCLGPKSM...</td>\n",
       "      <td>PB.1091.13</td>\n",
       "      <td>[PB.1091.13]</td>\n",
       "      <td>PB.1091.13</td>\n",
       "      <td>24</td>\n",
       "      <td>8.033045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MMGHRPVLVLSQNTKRESGRKVQSGNINAAKTIADIIRTCLGPKSM...</td>\n",
       "      <td>PB.1091.14</td>\n",
       "      <td>[PB.1091.14]</td>\n",
       "      <td>PB.1091.14</td>\n",
       "      <td>4</td>\n",
       "      <td>1.338841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>MVYMFQYDSTHGKFHGTVKAENGKLVINGNPITIFQERDPSKIKWG...</td>\n",
       "      <td>PB.9726.27|PB.9726.28</td>\n",
       "      <td>[PB.9726.27, PB.9726.28]</td>\n",
       "      <td>PB.9726.27</td>\n",
       "      <td>117</td>\n",
       "      <td>39.161096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>MEKAGAHLQGGAKRVIISAPSADAPMFVMGVNHEKYDNSLKIISNA...</td>\n",
       "      <td>PB.9726.29|PB.9726.31</td>\n",
       "      <td>[PB.9726.29, PB.9726.31]</td>\n",
       "      <td>PB.9726.29</td>\n",
       "      <td>81</td>\n",
       "      <td>27.111528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>MEEMCFGEATRMVWLPWVYGNLVSLNMVLSPSPPHPHRRDPSKIKW...</td>\n",
       "      <td>PB.9726.30</td>\n",
       "      <td>[PB.9726.30]</td>\n",
       "      <td>PB.9726.30</td>\n",
       "      <td>2</td>\n",
       "      <td>0.669420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>MFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLM...</td>\n",
       "      <td>PB.9726.32</td>\n",
       "      <td>[PB.9726.32]</td>\n",
       "      <td>PB.9726.32</td>\n",
       "      <td>35</td>\n",
       "      <td>11.714858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>MPSCLLSLRFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMF...</td>\n",
       "      <td>PB.9726.5</td>\n",
       "      <td>[PB.9726.5]</td>\n",
       "      <td>PB.9726.5</td>\n",
       "      <td>2</td>\n",
       "      <td>0.669420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>354 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      protein_sequence  \\\n",
       "0    MKELGIWEPLAVKLQTYKTAVETAVLLLRIDDIVSGHKKKGDDQSR...   \n",
       "1    MMLNIINSSITTKAISRWSSLACNIALDAVKMVQFEENGRKEIDIK...   \n",
       "2    MMKMLLDPMGGIVMTNDGNAILREIQVQHPAAKSMIEISRTQDEEV...   \n",
       "3    MMGHRPVLVLSQNTKRESGRKVQSGNINAAKTIADIIRTCLGPKSM...   \n",
       "4    MMGHRPVLVLSQNTKRESGRKVQSGNINAAKTIADIIRTCLGPKSM...   \n",
       "..                                                 ...   \n",
       "349  MVYMFQYDSTHGKFHGTVKAENGKLVINGNPITIFQERDPSKIKWG...   \n",
       "350  MEKAGAHLQGGAKRVIISAPSADAPMFVMGVNHEKYDNSLKIISNA...   \n",
       "351  MEEMCFGEATRMVWLPWVYGNLVSLNMVLSPSPPHPHRRDPSKIKW...   \n",
       "352  MFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLM...   \n",
       "353  MPSCLLSLRFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMF...   \n",
       "\n",
       "                                               pb_accs  \\\n",
       "0                                            PB.1091.1   \n",
       "1                                PB.1091.10|PB.1091.19   \n",
       "2    PB.1091.2|PB.1091.9|PB.1091.11|PB.1091.12|PB.1...   \n",
       "3                                           PB.1091.13   \n",
       "4                                           PB.1091.14   \n",
       "..                                                 ...   \n",
       "349                              PB.9726.27|PB.9726.28   \n",
       "350                              PB.9726.29|PB.9726.31   \n",
       "351                                         PB.9726.30   \n",
       "352                                         PB.9726.32   \n",
       "353                                          PB.9726.5   \n",
       "\n",
       "                                            accessions    base_acc   FL  \\\n",
       "0                                          [PB.1091.1]   PB.1091.1    2   \n",
       "1                             [PB.1091.10, PB.1091.19]  PB.1091.10   34   \n",
       "2    [PB.1091.2, PB.1091.9, PB.1091.11, PB.1091.12,...   PB.1091.2   51   \n",
       "3                                         [PB.1091.13]  PB.1091.13   24   \n",
       "4                                         [PB.1091.14]  PB.1091.14    4   \n",
       "..                                                 ...         ...  ...   \n",
       "349                           [PB.9726.27, PB.9726.28]  PB.9726.27  117   \n",
       "350                           [PB.9726.29, PB.9726.31]  PB.9726.29   81   \n",
       "351                                       [PB.9726.30]  PB.9726.30    2   \n",
       "352                                       [PB.9726.32]  PB.9726.32   35   \n",
       "353                                        [PB.9726.5]   PB.9726.5    2   \n",
       "\n",
       "           CPM  \n",
       "0     0.669420  \n",
       "1    11.380147  \n",
       "2    17.070221  \n",
       "3     8.033045  \n",
       "4     1.338841  \n",
       "..         ...  \n",
       "349  39.161096  \n",
       "350  27.111528  \n",
       "351   0.669420  \n",
       "352  11.714858  \n",
       "353   0.669420  \n",
       "\n",
       "[354 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pacbio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['transcript_id', 'gene_id']\n"
     ]
    }
   ],
   "source": [
    "sg = read_gtf('/mnt/shared/ubuntu/session_data/data/input/jurkat_corrected.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14361</td>\n",
       "      <td>29347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>14361</td>\n",
       "      <td>14829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>16765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>16858</td>\n",
       "      <td>17055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.1.1</td>\n",
       "      <td>PB.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606530</th>\n",
       "      <td>chr4</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>49305756</td>\n",
       "      <td>49308982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.16704.1</td>\n",
       "      <td>PB.16704.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606531</th>\n",
       "      <td>chr4</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>49510164</td>\n",
       "      <td>49512886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.16701.1</td>\n",
       "      <td>PB.16701.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606532</th>\n",
       "      <td>chr4</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>49510164</td>\n",
       "      <td>49512886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.16701.1</td>\n",
       "      <td>PB.16701.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606533</th>\n",
       "      <td>chr4</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>49510188</td>\n",
       "      <td>49511286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.16703.1_dup2</td>\n",
       "      <td>PB.16703.1_dup2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606534</th>\n",
       "      <td>chr4</td>\n",
       "      <td>hg38_canon</td>\n",
       "      <td>exon</td>\n",
       "      <td>49510188</td>\n",
       "      <td>49511286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>PB.16703.1_dup2</td>\n",
       "      <td>PB.16703.1_dup2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1606535 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname      source     feature     start       end  score strand  \\\n",
       "0          chr1  hg38_canon  transcript     14361     29347    NaN      -   \n",
       "1          chr1  hg38_canon        exon     14361     14829    NaN      -   \n",
       "2          chr1  hg38_canon        exon     14970     15038    NaN      -   \n",
       "3          chr1  hg38_canon        exon     15796     16765    NaN      -   \n",
       "4          chr1  hg38_canon        exon     16858     17055    NaN      -   \n",
       "...         ...         ...         ...       ...       ...    ...    ...   \n",
       "1606530    chr4  hg38_canon        exon  49305756  49308982    NaN      +   \n",
       "1606531    chr4  hg38_canon  transcript  49510164  49512886    NaN      -   \n",
       "1606532    chr4  hg38_canon        exon  49510164  49512886    NaN      -   \n",
       "1606533    chr4  hg38_canon  transcript  49510188  49511286    NaN      +   \n",
       "1606534    chr4  hg38_canon        exon  49510188  49511286    NaN      +   \n",
       "\n",
       "        frame    transcript_id          gene_id  \n",
       "0           0           PB.1.1           PB.1.1  \n",
       "1           0           PB.1.1           PB.1.1  \n",
       "2           0           PB.1.1           PB.1.1  \n",
       "3           0           PB.1.1           PB.1.1  \n",
       "4           0           PB.1.1           PB.1.1  \n",
       "...       ...              ...              ...  \n",
       "1606530     0       PB.16704.1       PB.16704.1  \n",
       "1606531     0       PB.16701.1       PB.16701.1  \n",
       "1606532     0       PB.16701.1       PB.16701.1  \n",
       "1606533     0  PB.16703.1_dup2  PB.16703.1_dup2  \n",
       "1606534     0  PB.16703.1_dup2  PB.16703.1_dup2  \n",
       "\n",
       "[1606535 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
